src/main/java/uk/gov/companieshouse/transactions/web/controller/confirmation/ConfirmationController.java

```diff
@@ -26,17 +30,38 @@
 
     private static final String ERROR_PAGE = "error";
 
+    private static final String PAYMENT_STATE = "payment_state";
+
     @Autowired
     private TransactionsService transactionsService;
 
     @Autowired
     private ConfirmationService confirmationService;
 
+    @Autowired
+    private SessionService sessionService;
+
     @GetMapping
     public String getConfirmation(@PathVariable String transactionId,
+                                  @RequestParam("state") Optional<String> paymentState,
                                   HttpServletRequest request,
                                   Model model) {
 
+        if (paymentState.isPresent()) {
+
+            Map<String, Object> sessionData = sessionService.getSessionDataFromContext();
+
+            String sessionPaymentState = (String) sessionData.get(PAYMENT_STATE);
+            sessionData.remove(PAYMENT_STATE);
+
+            if (!sessionPaymentState.equals(paymentState.get())) {
```

Comment:
Should we switch this line to read `if (!(paymentState.get()).equals(sessionPaymentState)) {` to get a `null` check on `sessionPaymentState` for free in case the Map doesn't contain the `payment_state` key for some reason?

Reply:
Good call! Flipped the equals check 👍